application/src/main/java/org/opentripplanner/transit/model/timetable/ScheduledTripTimes.java

```diff
@@ -185,6 +186,28 @@ public int sortIndex() {
     return getDepartureTime(0);
   }
 
+  @Override
+  public int firstScheduledDepartureTime() {
+    // trips that mix fixed times and flex windows can have negative departure times
+    for (var t : departureTimes) {
+      if (t > -1) {
+        return timeShifted(t);
+      }
+    }
+    return -1;
+  }
+
+  @Override
+  public int lastScheduledArrivalTime() {
+    // trips that mix fixed times and flex windows can have negative arrival times
+    for (var t : ArrayUtils.reversedCopy(arrivalTimes)) {
```

Comment:
I don't know if it matters at all, but why don't you just loop through this in reverse?

Reply:
I concentrated a bit and did it in https://github.com/opentripplanner/OpenTripPlanner/pull/6634/commits/3a6c93b6dcf223500d1a08304851c73174a73531